### ISSUE TITLE
Use of default calling convention on core 2.0 causes GC issues.

### ISSUE THREAD

**Issue by ThadHouse** [Firwood-Software/AdvancedDLSupport#82]
Something I noticed occasionally when testing this on .net core 2.0 was I would occasionally get gc crashes while in native calls. I noticed the restriction on using this in 32 bit mode on core 2.0, and that got me thinking and looking through the source code. Apparently, using the default managed calling convention does work on 64 bit, however it doesn't so the housekeeping unmanaged Calli does, which can lead to stability issues, especially with blocking and long running native calls (which I have several). I would recommend this functionality be removed, or at least hidden behind a flag that even 64 bit core 2.0 is unstable.

https://github.com/dotnet/coreclr/issues/19997

**Comment by Nihlus** [comment]
By default, ADL explicitly specifies Cdecl calling convention and not StdCall. As far as I understand this, it's only for StdCall?

**Comment by ThadHouse** [comment]
https://github.com/Firwood-Software/AdvanceDLSupport/blob/master/AdvancedDLSupport/ImplementationGenerators/Terminating/IndirectCallMethodImplementationGenerator.cs#L174

In the case where the runtime doesn't contain the UnmanagedCallingConvention overload for EmitCalli, it uses the regular managed calling convention. The managed calling convention works on x64, as its the same calling convention for both managed and unmanaged code, but it causes the GC issues.

```
Furthermore, when running under 32-bit .NET Core on Windows, calli is restricted to the __fastcall calling convention, and will ignore any hints to the contrary. This calling convention limitation is present on all .NET Core platforms, but is only relevant on x86. Calling conventions are, by design, ignored on other platforms in native code. The limitation in question is scheduled to be fixed in .NET Core 2.1.
```
This note in the doc is the same issue, however manifested in a different way. Only on 64 bit is it really an issue, but on 64 bit it can completely crash the runtime.

**Comment by Nihlus** [comment]
Ah, gotcha. In that case, simply throwing an exception when the correct overload isn't available might be the preferable option (along with documentiong why, of course).

**Comment by ThadHouse** [comment]
I think so. There are rare use cases where the managed calling convention could be used, but they need to be limited to code that does not take any locks and does not block. For a general purpose tool, its a good way to have things end very badly for people who don't know.

**Comment by Nihlus** [comment]
Yeah, I'd much prefer the "pit of success" over hard-to-debug errors popping up. I could maybe, maybe make it configurable, but eeeh...

**Comment by Nihlus** [comment]
What I've decided to do is bump ADL's required .NET standard version to 2.1 instead of 2.0 - this solves the issue of having to do janky patches in the code, and 2.1 is mature enough for me to move the library over.

**Comment by ThadHouse** [comment]
2.1 isn't released yet though, and blocks out framework completely. Plus 2.1 will require core 3 when it comes out in September.

**Comment by Nihlus** [comment]
Ah, damn, I thought it was released already. In that case, the exception option sounds more reasonable.

**Comment by Nihlus** [comment]
Are you having these problems on Windows or Linux, by the way?

**Comment by ThadHouse** [comment]
I was having them on Linux, but according to that issue posted above it can happen on any platform. The issue just stems from it doesn't do the function wrappers, so the GC assumes it's running managed code and can take full control, which when running native code it can't, so it gets confused. I've had both deadlocks and crashes occur.

**Comment by Nihlus** [comment]
Gotcha, thanks.

**Comment by Nihlus** [comment]
I've implemented the patch, bumped the version up to 3.0 since this is a breaking fix, and pushed the package to nuget.

**Comment by Nihlus** [comment]
We've had to rewind master because of a broken merge, so I'm reopening this issue. The 3.0 release patch was reverted in the process, and has not yet been reapplied to master.

**Comment by Nihlus** [comment]
Closing, since this was fixed in #90 but the issue was never closed.
